pkg/agent/memberlist/cluster_test.go

```diff
@@ -0,0 +1,626 @@
+	assertEgressSelectResult(fakeEgress2, false, true)
+	assertEgressSelectResult(fakeEgress1, false, false)
+
+	// Test creating Node with invalid IP.
+	fakeNode := &v1.Node{
+		ObjectMeta: metav1.ObjectMeta{Name: "fakeNode0"},
+		Status:     v1.NodeStatus{Addresses: []v1.NodeAddress{{Type: v1.NodeInternalIP, Address: "x"}}},
+	}
+	assert.NoError(t, createNode(fakeCluster.clientSet, fakeNode))
+	assertEgressSelectResult(fakeEgress2, false, true)
+	assertEgressSelectResult(fakeEgress1, false, false)
+
+	// Test deleting Node
+	deleteNode := func(node *v1.Node) {
+		err := fakeCluster.clientSet.CoreV1().Nodes().Delete(context.TODO(), node.Name, metav1.DeleteOptions{})
+		if err != nil {
+			t.Fatalf("Delete Node error: %v", err)
+		}
+		assert.NoError(t, wait.Poll(time.Millisecond*100, time.Second, func() (done bool, err error) {
+			newNode, _ := fakeCluster.cluster.nodeLister.Get(node.Name)
+			return nil == newNode, nil
+		}))
+	}
+	deleteNode(localNode)
+	assertEgressSelectResult(fakeEgress2, false, true)
+	assertEgressSelectResult(fakeEgress1, false, false)
+
+	// Test creating Node with valid IP.
+	fakeNode1 := &v1.Node{
+		ObjectMeta: metav1.ObjectMeta{Name: "fakeNode1"},
+		Status:     v1.NodeStatus{Addresses: []v1.NodeAddress{{Type: v1.NodeInternalIP, Address: "1.1.1.1"}}},
+	}
+	assert.NoError(t, createNode(fakeCluster.clientSet, fakeNode1))
+	assertEgressSelectResult(fakeEgress2, false, true)
+	assertEgressSelectResult(fakeEgress1, false, false)
+}
+
+func TestCluster_ShouldSelectNodeFailedOrAddedByConsistentHash(t *testing.T) {
+	const egressNum = 100
+	expectEgressSeqSum := func(n int) int {
+		count := 0
+		for i := 0; i < n; i++ {
+			count += i
+		}
+		return count
+	}(egressNum)
+
+	genNodes := func(n int) []string {
+		nodes := make([]string, n)
+		for i := 0; i < n; i++ {
+			nodes[i] = fmt.Sprintf("node-%d", i)
+		}
+		return nodes
+	}
+
+	nodes := genNodes(12)
+	testCases := []struct {
+		name           string
+		nodes          []string
+		consistentHash *consistenthash.Map
+	}{
+		{
+			name:           fmt.Sprintf("Assign owner Node for %d Egress", egressNum),
+			nodes:          nodes[:10],
+			consistentHash: consistenthash.New(defaultVirtualNodeReplicas, nil),
+		},
+		{
+			// Failover, when Node failed, Egress should move to available Node.
+			name:           "A Node fail then Egress should move",
+			nodes:          nodes[1:10],
+			consistentHash: consistenthash.New(defaultVirtualNodeReplicas, nil),
+		},
+		{
+			// Egress should move when Node added in cluster.
+			name:           "Add new Node then Egress should move",
+			nodes:          nodes[:11],
+			consistentHash: consistenthash.New(defaultVirtualNodeReplicas, nil),
+		},
+		{
+			name:           fmt.Sprintf("Recover to %d nodes", 10),
+			nodes:          nodes[:10],
+			consistentHash: consistenthash.New(defaultVirtualNodeReplicas, nil),
```

Comment:
I'm a bit confused by the test:
1. what does expectEgressSeqSum mean?
2. Does the 1st subtest have any difference from the 4th one given they have exactly same input?
3. What is it verifying given there is no expected value and it doesn't call any function/method in source file?

I feel it's likely just testing `consistentHash` directly which should be already covered by TestCluster_ShouldSelectEgress. If so, maybe just remove this one.

Reply:
expectEgressSeqSum is the sum of all selected Egress sequence numbers. It ensures each Egress has been selected once. That is, Egresses distributed in different Nodes, no duplicate or missing items. Maybe I should rephrase nodeSelectedForEgress and use ShouldSelectEgress in this test case.

Reply:
If verifying duplicate or missing items is the purpose, maybe you can just make a little change to`TestCluster_ShouldSelectEgress`;
```
genNodes := func(n int) []string {
	nodes := make([]string, n)
	for i := 0; i < n; i++ {
		nodes[i] = fmt.Sprintf("node-%d", i)
	}
	return nodes
}

fakeEIPName := "fakeExternalIPPool"
fakeEgress := &crdv1a2.Egress{
	ObjectMeta: metav1.ObjectMeta{Name: "fakeEgress"},
	Spec:       crdv1a2.EgressSpec{ExternalIPPool: fakeEIPName, EgressIP: tCase.egressIP},
}
consistentHashMap := newNodeConsistentHashMap()
consistentHashMap.Add(genNodes(tCase.nodeNum)...)

fakeCluster := &Cluster{
	consistentHashMap: map[string]*consistenthash.Map{fakeEIPName: consistentHashMap},
}

for i := 0; i < n; i++ {
	node := fmt.Sprintf("node-%d", i)
	fakeCluster.nodeName = node
	selected, err := fakeCluster.ShouldSelectEgress(fakeEgress)
	assert.NoError(t, err)
	assert.Equal(t, node == tCase.expectedNode, selected, "Selected Node for Egress not match")
}
```

Reply:
done